### PR TITLE
fix(dynamodb-table): dynamodb-table sync & reconcile

### LINKED ISSUE #580: Dynamodb writeConnectionSecretToRef does not write secret data
### What happened?

Creaded Dynamobd with
```yaml
    reason: Available
    status: "True"
    type: Ready

    reason: ReconcileSuccess
    status: "True"
    type: Synced
```
but no connection string written to the writeConnectionSecretToRef secret. Its created with no data?

Secret (redacted) :
```yaml
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: "2021-03-05T03:26:00Z"
  name: dynamodb-my-db
  ownerReferences:
  - apiVersion: dynamodb.aws.crossplane.io/v1alpha1
    controller: true
    kind: Table
type: connection.crossplane.io/v1alpha1
```

### How can we reproduce it?

```yaml
apiVersion: dynamodb.aws.crossplane.io/v1alpha1
kind: Table
#...#
  forProvider:
    attributeDefinitions:
    - attributeName: attribute1
      attributeType: S
    keySchema:
    - attributeName: attribute1
      keyType: HASH
    provisionedThroughput:
      readCapacityUnits: 1
      writeCapacityUnits: 1
    region: ap-southeast-2
    streamSpecification:
      streamEnabled: true
      streamViewType: NEW_AND_OLD_IMAGES
  providerConfigRef:
    name: default
  writeConnectionSecretToRef:
    name: dynamodb-my-db
    namespace: #...#
```

### What environment did it happen in?
Crossplane version: v0.17.0

### PR DESCRIPTION
**Pull request by haarchri**
Signed-off-by: Christopher Haar <chhaar30@googlemail.com>

### Description of your changes

- implemented postUpdate
- implemented connectionSecret
- implemented in preUpdate SSESpecification, BillingMode
- implemented for SSESpecification lateInit to get kms keyarn

Fixes #580 #637

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
added new examples from #580 and tested create, update & deletion

```
kubectl get tables
NAME                       READY   SYNCED   EXTERNAL-NAME
sample-table-ppr           True    True     sample-table-ppr
sample-table-with-sse      True    True     sample-table-with-sse
sample-table-with-stream   True    True     sample-table-with-stream
sample-table-without-sse   True    True     sample-table-without-sse
```

image

connection secret is now possible: for LatestStreamArn, LatestStreamLabel, TableArn, TableName

stream_enabled = true
```
kubectl get secret sample-table-with-stream -n crossplane-system -o yaml
apiVersion: v1
data:
  LatestStreamArn: YXJuOmF3czpkeW5hbW9kYjpldS1jZW50cmFsLTE6MjU1OTMyNjQyOTI3OnRhYmxlL3NhbXBsZS10YWJsZS13aXRoLXN0cmVhbS9zdHJlYW0vMjAyMS0wOS0yMVQwODowMjoxMy4yNjU=
  LatestStreamLabel: MjAyMS0wOS0yMVQwODowMjoxMy4yNjU=
  TableArn: YXJuOmF3czpkeW5hbW9kYjpldS1jZW50cmFsLTE6MjU1OTMyNjQyOTI3OnRhYmxlL3NhbXBsZS10YWJsZS13aXRoLXN0cmVhbQ==
  TableName: c2FtcGxlLXRhYmxlLXdpdGgtc3RyZWFt
kind: Secret
...
```

stream_enabled = false
```
kubectl get secret sample-table-with-sse -n crossplane-system -o yaml
apiVersion: v1
data:
  LatestStreamArn: ""
  LatestStreamLabel: ""
  TableArn: YXJuOmF3czpkeW5hbW9kYjpldS1jZW50cmFsLTE6MjU1OTMyNjQyOTI3OnRhYmxlL3NhbXBsZS10YWJsZS13aXRoLXNzZQ==
  TableName: c2FtcGxlLXRhYmxlLXdpdGgtc3Nl
kind: Secret
```

### CODE REVIEW

**Comment by AaronME** [comment]
Might be some overlap here with https://github.com/crossplane/provider-aws/pull/493

**Comment by haarchri** [comment]
looks for me that we not really have a overlap here is table the other PR is related with backup or ?

File: pkg/controller/dynamodb/table/hooks.go [hunk 1]
```diff
@@ -308,12 +328,33 @@ func (e *updateClient) preUpdate(_ context.Context, cr *svcapitypes.Table, u *sv
...
 	*u = *newUpdateObj
+
 	return nil
 }
+
+func postUpdate(_ context.Context, cr *svcapitypes.Table, t *svcsdk.UpdateTableOutput, upd managed.ExternalUpdate, err error) (managed.ExternalUpdate, error) {
+	if err != nil {
+		// no update required - newUpdateObj was empty
+		// no enum available for this skip
+		if strings.Contains(err.Error(), "ProvisionedThroughput, BillingMode, UpdateStreamEnabled, GlobalSecondaryIndexUpdates or SSESpecification or ReplicaUpdates is required") {
```
Comment (chlunde) [inline pkg/controller/dynamodb/table/hooks.go:354]: Would a normal user hit this? is it possible to do something about this in `isUpToDate`?

Comment (haarchri) [reply pkg/controller/dynamodb/table/hooks.go:354]: @chlunde at the moment we have preUpdate with switch/case and creating one specific Update object with one of the mentioned keys - if nothing is to-do we receive this Message that we need to send one update  for the mentioned key- If you have a better way via IsUpToDate please make a suggestion

**Comment by jbw976** [comment]
This is great for you to provide a fix for this issue @haarchri! what all is left for getting this PR merged? I have heard/seen a few more folks running into these issues that you are addressing :)

File: pkg/controller/dynamodb/table/hooks.go (hunk 1 above)
Comment (haarchri) [reply pkg/controller/dynamodb/table/hooks.go:354]: @chlunde Any ideas for isUpToDate ? Or can we Go with the PR solution ?

**Review by negz — APPROVED** [review]

**Comment by negz** [comment]
Thanks for fixing this @haarchri! We've been hearing from a lot of folks that they'd like this fixed, so I'm going to move forward with it. I might open a small followup PR for some nits (e.g. error matching and connection secret keys).

**Pull request merged**